Add Event Study (aka Dynamic Difference in Differences) functionality

drbenvincent opened the pull request:
* Towards #384 

This pull request adds support for event study analysis to the CausalPy package. The main changes include introducing the new `EventStudy` class, updating the package exports to include it, and providing a utility function for generating synthetic panel data suitable for event study and dynamic DiD analyses.

- [x] This implementation does not deal with staggered treatments. Add a warning callout box, as data validation with exception
- [x] Update model description - the model being used is not static, it depends on the user provided formula. Give some examples. 
- [x] Add an example where we remove the time fixed effects and add in other predictors or some form of seasonal trend. 
- [x] Check the integration with the reporting layer. 
- [ ] ~~Consider adding the time relative to treatment column in the notebook, not hidden in the experiment class.~~ 
- [x] Ensure we are exposing plot options via arguments
- [x] Simplify data plot using a 1-liner with seaborn
- [x] Why is the event study column not appearing in the table on home page?

---

⚠️ Includes a more widespread big fix identified by bugbot here https://github.com/pymc-labs/CausalPy/pull/584#discussion_r2595103978 which will affect multiple experiment classes.

<!-- readthedocs-preview causalpy start -->
----
📚 Documentation preview 📚: https://causalpy--584.org.readthedocs.build/en/584/

<!-- readthedocs-preview causalpy end -->

drbenvincent committed:
initial commit

drbenvincent committed:
add new notebook to index so it renders in the docs

drbenvincent committed:
Add notebook indexing guideline to documentation

drbenvincent committed:
add cell tags

drbenvincent committed:
add glossary entry + refs

drbenvincent committed:
update README with event study functionality notes

drbenvincent committed:
fix math formatting

drbenvincent committed:
add reference

drbenvincent committed:
Refactor EventStudy to use patsy formula for FEs

The EventStudy class now requires a patsy-style formula to specify the outcome and fixed effects, removing the separate outcome_col argument. Design matrix construction uses patsy, and event-time dummies are appended. Input validation checks for formula presence, and tests and documentation are updated to reflect the new API and output format.

drbenvincent committed:
Clarify EventStudy formula and event-time dummies

Expanded documentation to explain the patsy formula syntax, the role of unit and time fixed effects, and how event-time dummies ($\beta_k$) are automatically constructed by the EventStudy class. Added details on the event window and reference event time parameters for clearer guidance.

drbenvincent committed:
Add check and warning for staggered adoption in EventStudy

Added a warning in the EventStudy class and documentation that the implementation only supports simultaneous treatment timing and does not support staggered adoption. Introduced a validation to raise a DataException if treated units have different treatment times. Added a corresponding test to ensure staggered adoption raises an error, and updated the notebook to clarify estimator limitations.

drbenvincent committed:
Add plot customization options to EventStudy

Enhanced the _bayesian_plot and _ols_plot methods in EventStudy to support configurable figure size and HDI probability. Updated docstrings to document new parameters and improved plot labeling for clarity.

drbenvincent committed:
plot data with seaborn, not matplotlib

drbenvincent committed:
re-render notebook

drbenvincent committed:
run make uml

drbenvincent committed:
Add event study effect summary and reporting

Introduces event study support to effect_summary(), including parallel trends check and dynamic effect reporting. Updates event study class to allow HDI probability customization and reporting, and extends documentation with effect summary usage and interpretation.

drbenvincent committed:
Add more model examples to notebook + add time-varying predictors

The `generate_event_study_data` function now supports optional time-varying predictors generated as AR(1) processes, controlled by new parameters: `predictor_effects`, `ar_phi`, and `ar_scale`. Added the `generate_ar1_series` utility function. Updated docstrings and examples to reflect these changes. The event study PyMC notebook was updated with additional analysis and improved section headings.

drbenvincent committed:
minor changes + re-run notebook

drbenvincent committed:
update the homepage (index.md, not README.md)

drbenvincent committed:
Add note about dual README files in documentation

drbenvincent committed:
Add tests for EventStudy effect_summary method

Introduces integration tests for the EventStudy.effect_summary method using both PyMC and sklearn models. Tests verify the returned EffectSummary object, its table and text attributes, and key output elements.

drbenvincent committed:
re-run notebook

drbenvincent committed:
minor notebook edit

juanitorduz commented:
I will take a look in the next few days :)

nialloulton commented:
@cursor review

drbenvincent committed:
fix spelling error

drbenvincent committed:
Document Event Study support in effect_summary

Added Event Study to the experiment support table in reporting_statistics.md and updated AGENTS.md to instruct updating the table when adding new experiment types.

drbenvincent committed:
Add comprehensive tests for reporting and synthetic data modules

This commit adds extensive new tests to test_reporting.py and test_synthetic_data.py, covering error handling, experiment type detection, OLS statistics edge cases, prose and table generation for various models, and all synthetic data generation utilities. These tests improve coverage and robustness for reporting and data simulation functions.

drbenvincent committed:
Clarify indicator function and dummy variable usage in event study

Expanded documentation in both the EventStudy class and the event study PyMC notebook to explain the equivalence between indicator functions and dummy variables. Added details on how dummy variables are constructed for each event time, the omission of the reference period to avoid multicollinearity, and the interpretation of regression coefficients as ATT at each event time.

juanitorduz commented on causalpy/tests/test_event_study.py:
```diff
@@ -0,0 +1,542 @@
+
+
+# ============================================================================
+# Integration Tests with PyMC
```
@drbenvincent, can we add a cursor to add tests for each individual method of the `EventStudy` class? This could be especially helpful to detect edge cases (say nans and stuff like that)

juanitorduz reviewed (changes requested):
Very cool! This is super helpful in marketing (the effect of a marketing campaign depends on time!). We should try to find other real datasets to showcase this.

Minor comments. 

Suggestion -> add more tests for the indidual methods and address @bugbot comments

drbenvincent committed:
Fix control unit detection in EventStudy for np.inf and np.nan

Control units marked with np.inf or np.nan in the treat_time column are now correctly excluded from treated unit checks in EventStudy. Adds tests to ensure control units with np.inf, np.nan, or mixed markers are handled properly, preventing false staggered adoption errors.

drbenvincent committed:
Refactor event study summary and update tests

Improved rounding logic in EventStudy.get_event_time_summary for PyMC and sklearn models, ensuring consistent application of the round_to parameter. Refactored and expanded test_event_study.py to cover more edge cases, input validation, and integration scenarios, including new tests for rounding, event window handling, and control units. Cleaned up and reorganized tests for clarity and maintainability.

drbenvincent committed:
Merge branch 'main' into event-study

drbenvincent commented on causalpy/experiments/event_study.py:
```diff
@@ -0,0 +1,636 @@
+                }
+            rows.append(row)
+
+        return pd.DataFrame(rows)
```
Should be resolved by https://github.com/pymc-labs/CausalPy/pull/584/commits/364faed3b43222af8552d4507afbe531d921cd7a I think

drbenvincent committed:
make uml

drbenvincent committed:
Handle NaN values by filtering rows in experiment classes

Updates all experiment classes to filter out rows with NaN values after patsy design matrix construction, ensuring consistent shapes between data and design matrices. Adds comprehensive tests for NaN handling across all experiment classes to verify correct filtering and prevent shape mismatch errors.

drbenvincent committed:
Add warnings for panel data edge cases in EventStudy

Enhanced the EventStudy class to emit warnings for unbalanced panels, gaps in time periods, and excessive data loss due to NaN filtering. Added comprehensive unit tests to cover these edge cases and verify warning behavior. Updated interrogate badge to reflect increased coverage.

drbenvincent commented on causalpy/experiments/event_study.py:
```diff
@@ -0,0 +1,648 @@
+            ).astype(float)
+
+        # Combine patsy design matrix with event-time dummies
+        X_df = pd.concat([X_df, event_time_dummies], axis=1)
```
Good idea, but a) not sure if we need an expternal package for de-meaning? b) let's have this as an iteration on top of the first mvp (i.e. yes, but not in this PR)

drbenvincent committed:
Cite some classic papers

drbenvincent commented:
<p>Done, along with some other citations</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/584/discussion/#comment-3655260502'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

drbenvincent committed:
Update pip commands to use python -m pip in docs

Replaced direct 'pip' calls with 'python -m pip' in CONTRIBUTING.md to ensure the correct pip is used within the conda environment. Added a note explaining the reason for this change.

drbenvincent committed:
Merge branch 'main' into event-study

drbenvincent commented on causalpy/experiments/event_study.py:
```diff
@@ -0,0 +1,714 @@
+    ... )
+    >>> result = cp.EventStudy(
+    ...     df,
+    ...     formula="y ~ C(unit) + C(time)",
```
Sorry for the AI evaluation, but some major concerns with that proposal

--- 

## The Suggestion

Use patsy's categorical handling to simplify the event-time dummy creation:

```python
"y ~ C(unit) + C(rel_time, Treatment(reference=-1))"
```

With pre-processing to compute `rel_time` before calling patsy.

## Benefits

- Simpler code (removes ~50 lines of manual dummy creation)
- More idiomatic use of patsy
- Reference category handled automatically via `Treatment(reference=-1)`

## Concerns

### 1. Control Units Would Be Dropped

This is the most significant issue. Control units have `NaN` for treatment time, so `rel_time` would be `NaN`. Patsy drops rows with `NaN` values entirely.

**Current behavior**: Control units are kept in the estimation. Their event-time dummies are all 0, but they contribute to the time fixed effects estimation.

**Proposed behavior**: Control units would be excluded, fundamentally changing the model and breaking the parallel trends comparison.

### 2. Loss of Calendar Time Fixed Effects

The proposed formula replaces calendar time FEs with event-time dummies. But a proper TWFE event study needs **both**:

```
Y_it = α_i + λ_t + Σ β_k · 1{E_it = k} + ε_it
        ↑      ↑         ↑
      unit   time    event-time
```

With only `C(rel_time)`, we lose `λ_t` (calendar time effects). This changes the identifying assumptions of the model.

### 3. Event Window Handling

Currently, only event times within the window (e.g., -5 to +5) get dummies. Observations outside the window are kept but don't contribute to event-time indicators.

With patsy handling everything, we'd need to either:
- Filter out observations with `rel_time` outside the window (loses data)
- Let patsy create dummies for ALL unique `rel_time` values (different model, many more parameters)
- Bin/clip values outside the window to special categories

### 4. Minor Issues

- **Coefficient naming**: Changes from `event_time_0` to `C(rel_time, Treatment(reference=-1))[T.0]`
- **Dynamic reference**: Would need to build formula strings dynamically to inject `reference_event_time`
- **Test updates**: Existing tests check current coefficient naming

## Possible Middle Ground

Keep the current model structure (which correctly handles control units and maintains both time FEs and event-time dummies) but simplify the internal implementation where possible.

## Questions

1. Am I missing something that would address the control unit issue?
2. Is the loss of calendar time FEs intentional in your suggestion, or should both be included?
3. Any thoughts on handling the event window bounds?

Happy to implement if we can resolve these concerns!

drbenvincent committed:
Merge branch 'main' into event-study

Resolve conflicts by combining main's refactored experiment structure
(_build_design_matrices, _prepare_data, algorithm methods, default model
classes) with event-study's NaN handling (return_type="dataframe" in
dmatrices calls and index-based row filtering).

Conflicting files:
- README.md: keep both new method rows (Event Study + Staggered DiD)
- 6 experiment files: take main's refactored methods, add PR's NaN handling
- interrogate_badge.svg: accept main's deletion
- references.bib: keep all new references from both sides

Made-with: Cursor

drbenvincent committed:
Fix post-merge issues: restore lost function and update deprecated freq

- Restore _compute_statistics_rkink_ols stub that was dropped during
  auto-merge of reporting.py
- Update pandas frequency string 'M' -> 'ME' in simulate_data.py
  (deprecated since pandas 2.2)

Made-with: Cursor

drbenvincent committed:
Merge remote-tracking branch 'origin/main' into event-study

Made-with: Cursor

drbenvincent committed:
Restore legacy reporting helper names for event-study branch

Made-with: Cursor

NathanielF reviewed (approved):
Sorry it took so long.

drbenvincent commented:
## Automated triage

**Recommendation:** `review:high` — no decision gate identified.

**Why:**
- Major new feature: `EventStudy` class (759 lines), new `reporting.py` logic (201 lines), new data simulation functions.
- Cross-cutting: changes to `base.py`, `reporting.py`, multiple experiment files, glossary, and references.
- Large test suite: `test_event_study.py` (1087 lines), `test_nan_handling.py`, `test_reporting.py`, `test_synthetic_data.py`.
- All CI checks pass. PR is conflicted (DIRTY). Review already has requested changes. Includes a bug fix affecting multiple experiment classes.

**Review focus:**
1. Verify the event study model handles non-staggered treatment correctly (the PR notes this limitation with a warning).
2. Review the bug fix that affects multiple experiment classes (noted in the PR description).
3. Confirm the reporting layer integration works for event study-specific outputs.

**Confidence:** high

drbenvincent commented:
@juanitorduz — when you have a moment, could you take another look at this PR on the current head (`95bbfc5c`)?

Your Dec review asked for more per-method tests and bugbot fixes. Since then we've:

- merged latest `main` and resolved conflicts
- added extensive `EventStudy` tests plus `test_nan_handling.py` across experiment classes
- fixed the bugbot items (`np.inf` controls, `round_to` in `get_event_time_summary`, NaN row/index alignment via `build_formula_matrices` + index filtering)
- added unified `plot()` on `EventStudy` and gallery notebook `event-study-pymc`

CI is green on the latest push. Deferred items from the thread (real marketing datasets, demeaning FE for large N, patsy `Treatment()` refactor) are noted in the updated PR description as follow-ups.

If you're happy with the current state, an approval would unblock merge. If anything still feels like a major sticking point, let us know and we'll address it.
